File: Day #453/Solution.java
// Day 453: Two nodes in a BST summing to K.
// Inorder -> sorted list, then two-pointer. Time O(n), Space O(n).
import java.util.ArrayList;
import java.util.List;

public class Solution {
    static class Node {
        int val;
        Node left, right;
        Node(int v) { val = v; }
    }

    static void inorder(Node n, List<Integer> out) {
        if (n == null) return;
        inorder(n.left, out);
        out.add(n.val);
        inorder(n.right, out);
    }

    static int[] twoSum(Node root, int k) {
        List<Integer> a = new ArrayList<>();
        inorder(root, a);
        int i = 0, j = a.size() - 1;
        while (i < j) {
            int s = a.get(i) + a.get(j);
            if (s == k) return new int[]{a.get(i), a.get(j)};
            if (s < k) i++; else j--;
        }
        return new int[]{-1, -1};
    }

    public static void main(String[] args) {
        Node root = new Node(10);
        root.left = new Node(5);
        root.right = new Node(15);
        root.right.left = new Node(11);
        root.right.right = new Node(15);
        int[] r = twoSum(root, 20);
        System.out.println(r[0] + " and " + r[1]); // 5 and 15
    }
}

File: Day #453/solution.js
// Day 453: Two nodes in a BST summing to K.
// Inorder -> sorted array, then two-pointer. Time O(n), Space O(n).
class Node {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function inorder(n, out) {
  if (!n) return;
  inorder(n.left, out);
  out.push(n.val);
  inorder(n.right, out);
}

function twoSum(root, k) {
  const a = [];
  inorder(root, a);
  let i = 0, j = a.length - 1;
  while (i < j) {
    const s = a[i] + a[j];
    if (s === k) return [a[i], a[j]];
    if (s < k) i++; else j--;
  }
  return null;
}

const root = new Node(10, new Node(5), new Node(15, new Node(11), new Node(15)));
const [x, y] = twoSum(root, 20);
console.log(`${x} and ${y}`); // 5 and 15
